agent/ui_endpoint_test.go

```diff
@@ -1397,34 +1398,59 @@ func TestUIServiceTopology(t *testing.T) {
 		}
 	}
 
-	t.Run("request without kind", func(t *testing.T) {
-		req, _ := http.NewRequest("GET", "/v1/internal/ui/service-topology/ingress", nil)
-		resp := httptest.NewRecorder()
-		obj, err := a.srv.UIServiceTopology(resp, req)
-		require.Nil(t, err)
-		require.Nil(t, obj)
-		require.Equal(t, "Missing service kind", resp.Body.String())
-	})
-
-	t.Run("request with unsupported kind", func(t *testing.T) {
-		req, _ := http.NewRequest("GET", "/v1/internal/ui/service-topology/ingress?kind=not-a-kind", nil)
-		resp := httptest.NewRecorder()
-		obj, err := a.srv.UIServiceTopology(resp, req)
-		require.Nil(t, err)
-		require.Nil(t, obj)
-		require.Equal(t, `Unsupported service kind "not-a-kind"`, resp.Body.String())
-	})
+	type testCase struct {
+		name    string
+		httpReq *http.Request
+		want    *ServiceTopology
+		wantErr string
+	}
 
-	t.Run("ingress", func(t *testing.T) {
+	run := func(t *testing.T, tc testCase) {
```

Comment:
Hey @freddygv , this layout for table tests is the same one I use. I wrote up a short explanation of it here: https://github.com/hashicorp/consul/pull/10814#discussion_r686221701

I quite like this style because I think it allows the test to be read easily from top-to-bottom, instead of having to scroll back and forth to match the test logic with the test cases.

Reply:
I tried following @dnephin's pattern of putting the test logic closer to the testcase type (described [here](https://github.com/hashicorp/consul/pull/10814#discussion_r686221701)), which I thought was neat and legible. I can move this back down if you don't agree or would like to discuss the style elsewhere

Reply:
Oh woops dan was faster than me :D